aws/signer/v4/v4.go

```diff
@@ -0,0 +1,537 @@
+			// If the request is already signed, and the credentials have not
+			// expired, and the request is not too old ignore the signing request.
+			return r.URL.String(), ctx.SignedHeaderVals, nil
+		}
+		ctx.Time = time.Now()
+
+		// The credentials have expired for this request. The current signing
+		// is invalid, and needs to be request because the request will fail.
+		ctx.removePresign()
+		// Update the request's query string to ensure the values stays in
+		// sync in the case retrieving the new credentials fails.
+		ctx.Request.URL.RawQuery = ctx.Query.Encode()
+	}
+
+	var err error
+	ctx.credValues, err = v4.Credentials.Get()
+	if err != nil {
+		return "", http.Header{}, err
+	}
+
+	ctx.Query.Set("X-Amz-Algorithm", authHeaderPrefix)
+	if ctx.credValues.SessionToken != "" {
+		ctx.Query.Set("X-Amz-Security-Token", ctx.credValues.SessionToken)
+	} else {
+		ctx.Query.Del("X-Amz-Security-Token")
+	}
+
+	v4.build(ctx)
+
+	if v4.Debug.Matches(aws.LogDebugWithSigning) {
+		v4.logSigningInfo(ctx)
+	}
+
+	return r.URL.String(), ctx.SignedHeaderVals, nil
+}
+
+// Sign requests with signature version 4.
+//
+// Will sign the requests with the service config's Credentials object
+// Signing is skipped if the credentials is the credentials.AnonymousCredentials
+// object.
+func Sign(req *request.Request) {
+	// If the request does not need to be signed ignore the signing of the
+	// request if the AnonymousCredentials object is used.
+	if req.Config.Credentials == credentials.AnonymousCredentials {
+		return
+	}
+
+	region := req.ClientInfo.SigningRegion
+	if region == "" {
+		region = aws.StringValue(req.Config.Region)
+	}
+
+	name := req.ClientInfo.SigningName
+	if name == "" {
+		name = req.ClientInfo.ServiceName
+	}
+
+	v4 := NewSigner(req.Config.Credentials, func(v4 *Signer) {
+		v4.Debug = req.Config.LogLevel.Value()
+		v4.Logger = req.Config.Logger
+		v4.NotHoist = req.NotHoist
+	})
+
+	if req.ExpireTime != 0 {
+		_, signedHeaders, err := v4.presignWithBody(req.HTTPRequest, req.Body, name, region, req.ExpireTime, req.Time)
+		req.SignedHeaderVals = signedHeaders
+		req.Error = err
+	} else {
+		signedHeaders, err := v4.signWithBody(req.HTTPRequest, req.Body, name, region, req.Time)
+		req.SignedHeaderVals = signedHeaders
+		req.Error = err
+	}
+}
+
+const logSignInfoMsg = `DEBUG: Request Signiture:
+---[ CANONICAL STRING  ]-----------------------------
+%s
+---[ STRING TO SIGN ]--------------------------------
+%s%s
+-----------------------------------------------------`
+const logSignedURLMsg = `
+---[ SIGNED URL ]------------------------------------
+%s`
+
+func (v4 *Signer) logSigningInfo(ctx *signingCtx) {
+	signedURLMsg := ""
+	if ctx.isPresign {
+		signedURLMsg = fmt.Sprintf(logSignedURLMsg, ctx.Request.URL.String())
+	}
+	msg := fmt.Sprintf(logSignInfoMsg, ctx.canonicalString, ctx.stringToSign, signedURLMsg)
+	v4.Logger.Log(msg)
+}
+
+func (v4 *Signer) build(ctx *signingCtx) {
```

Comment:
What do you think about putting the `build` method on the `signingCtx` type instead of `Signer`. It doesn't look like any `Signer` value is being used here other than `NotHoist`. This could be passed in as a configuration param instead.

Alternatively we could change `build` to `sign` and keep the method on `Signer`.